### ISSUE TITLE
Bug: File ignored because outside of base path

### ISSUE THREAD

**Issue by Peredery** [eslint/eslint#19118]
### Environment

eslint version: v9.14.0

### What parser are you using?

Default (Espree)

### What did you do?

I have config file inside: /app/eslint.config.mjs and want to lint files inside /tmp/example.js

root@container:/app  npx eslint --flag unstable_config_lookup_from_file -c "./eslint.config.mjs" /tmp/example.js

```
/tmp/example.js
  0:0  warning  File ignored because outside of base path
```

### What did you expect to happen?

behaviour as before version 9.0, should check the file

### What actually happened?

 File ignored because outside of base path

### Link to Minimal Reproducible Example

https://github.com/Peredery/eslint_bug_report_-File-ignored-because-outside-of-base-path

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Tanujkanti4441** [comment]
Hi @Peredery, thanks for the issue, but i am not able to reproduce the error, can you please mention from which directory are you running ` npx eslint --flag unstable_config_lookup_from_file -c "./eslint.config.mjs" /tmp/outer_example.js` command.

**Comment by nrmmota** [comment]
I ran into the same issue on 9.16.0

If I run the eslint command inside one of the folders with the same base folder of the folder I want to test, it triggers this error.
cd /src/folder/lib/lint
eslint /src/folder/lib/code/*
 0:0  warning  File ignored because outside of base path

If I go / then run the eslint with full path for config and test dir it works fine.
cd /
eslint /src/folder/lib/lint/eslint.config.js /src/folder/lib/code/*

hope it helps.

**Comment by oo10** [comment]
I also encountered this.
When using previous versions of ESLint in nodejs, we were able to lint files located outside the nodejs runtime directory using lintFiles. However, after upgrading to ESLint 9, it seems to no longer work. The output now shows a warning: "File ignored because outside of base path".

**Comment by nzakas** [comment]
@eslint/eslint-tsc I could use some feedback on my comment here: https://github.com/eslint/eslint/issues/19118#issuecomment-2518650141

**Comment by fasttime** [comment]
> So, I think this brings up an interesting question: If a flat config file is using something like `"**/*.js"`, should that match a file that is outside of the base path?
> 
> I can see a strong argument for "yes", in that you're stating all JavaScript files should be matched. That would allow the repro case to work as expected because it's just using the default `files` entries from ESLint.

I also think that if a user writes a files pattern like `"**/*.js"` in a config file, it's because they probably intend to lint JS files in the parent directory of the config file's location. So I agree that there's an argument for matching files outside the base path.

A complication with the current config system is that file patterns are not relative to a config file's location, but to the current directory. But regardless, it should be possible to match files outside the base path even for complex patterns with the proper logic.

A non-breaking alternative would be adding a `basePath` property to configs as suggested in [this comment](https://github.com/eslint/eslint/issues/18948#issuecomment-2564809279). This would allow users to specify the relative location of their files while still excluding files outside the base path.

**Comment by nzakas** [comment]
> A complication with the current config system is that file patterns are not relative to a config file's location, but to the current directory. But regardless, it should be possible to match files outside the base path even for complex patterns with the proper logic.

I think this is backwards? By default, patterns are matched relative to the config file location. Only when `-c` is used do we match from the current working directory. Or am I missing something?

> A non-breaking alternative would be adding a `basePath` property to configs as suggested in [this comment](https://github.com/eslint/eslint/issues/18948#issuecomment-2564809279). This would allow users to specify the relative location of their files while still excluding files outside the base path.

I'm not sure adding a `basePath` property to configs solves the same problem. If `**/*.js` is meant to mean "any JavaScript file", then to mimic that behavior using `basePath`, you'd have to add multiple config objects each with a `basePath` for any directory you want the config file to work for. 

We could put the new `**/*.js` behavior behind a flag to let people opt-in to avoid a breaking change (`--flag v10_file_matching`).

**Comment by mdjermanovic** [comment]
How would we determine which patterns should apply when matching files that are outside the base path? Would that be all patterns that start with `**/`?

**Comment by nzakas** [comment]
> How would we determine which patterns should apply when matching files that are outside the base path? Would that be all patterns that start with `**/`?

That's what I was thinking. So if we see `**/` at the start of a pattern, we just don't apply the base path before matching.

**Comment by slolo2000** [comment]
Hello,

I have migrated from v8.57.1 to v9.17.0 and I encountered the same problem with the following error: "File ignored because outside of base path."
My case is slightly different because I use it with "grunt js" but I try to lint code outside of the main program.

Otherwise, I can reproduce also the bug mentioned by @Peredery.

Is it a real bug or is there a way to get around the problem?

**Comment by nzakas** [comment]
@slolo2000 This is working as designed and is not a bug. We are discussing whether we want to make a change and what change that will be.

**Comment by mdjermanovic** [comment]
> > How would we determine which patterns should apply when matching files that are outside the base path? Would that be all patterns that start with `**/`?
> 
> That's what I was thinking. So if we see `**/` at the start of a pattern, we just don't apply the base path before matching.

Not applying the base path means that we'll test if the pattern matches absolute paths instead of relative paths?

**Comment by fasttime** [comment]
> > A complication with the current config system is that file patterns are not relative to a config file's location, but to the current directory. But regardless, it should be possible to match files outside the base path even for complex patterns with the proper logic.
> 
> I think this is backwards? By default, patterns are matched relative to the config file location. Only when `-c` is used do we match from the current working directory. Or am I missing something?

You are correct. I only considered the case where the config path is explicitly specified with `-c` like in the example in this issue.

> > A non-breaking alternative would be adding a `basePath` property to configs as suggested in [this comment](https://github.com/eslint/eslint/issues/18948#issuecomment-2564809279). This would allow users to specify the relative location of their files while still excluding files outside the base path.
> 
> I'm not sure adding a `basePath` property to configs solves the same problem. If `**/*.js` is meant to mean "any JavaScript file", then to mimic that behavior using `basePath`, you'd have to add multiple config objects each with a `basePath` for any directory you want the config file to work for.

One could also set `basePath` to a parent directory such as `".."` and have the directories to be linted included in `files` patterns, e.g. `["lib/**/*.js", "src/**/*.js"]`. This would allow matching JS files in `../lib` or `../src` relative to the config's location, whereas files patterns starting with `../` would be still considered external.

> We could put the new `**/*.js` behavior behind a flag to let people opt-in to avoid a breaking change (`--flag v10_file_matching`).

My idea was to add a new optional public key `basePath` to config objects. That would not be a breaking change even behind a feature flag.

**Comment by idhard** [comment]
is there a workaround for this issue until the flag is available ? i had to migrate to 9.x and i'm facing same issue after using flat configs.

**Comment by nzakas** [comment]
@idhard you'll need to either use `-c` to specify a config file explicitly or change the location from which you run ESLint.

**Comment by nzakas** [comment]
@eslint/eslint-tsc shall we make `**/`-prefixed paths match all files regardless of base path?

**Comment by idhard** [comment]
@nzakas thanks for the tip, unfortunately i'm initializing Eslint programmatically  as part of a custom build tool. I'm not able to specify where the CLI is called from. Your suggestion to change how eslint resolves `**/` seems a good idea to me.

**Comment by fasttime** [comment]
> [@eslint/eslint-tsc](https://github.com/orgs/eslint/teams/eslint-tsc) shall we make `**/`-prefixed paths match all files regardless of base path?

I have some questions about this proposal:
* (How) will it be possible to match only files in the base path? That is, do what a pattern like `**/` is currently doing?
* (How) will it be possible to match specific files or directories outside the base path?

Both can be currently achieved by adding a config file to the parent directory and running ESLint from there, but I'm assuming that the user cannot or doesn't want to change the current directory (or `basePath`)?

**Comment by nzakas** [comment]
> * (How) will it be possible to match only files in the base path? That is, do what a pattern like `**/` is currently doing?

They could do `/**/*.js`. 

> * (How) will it be possible to match specific files or directories outside the base path?

I'd say "no" to this. Unless we want to start supporting patterns beginning with `..`, but I don't think we're at the point where there's any evidence we need this.

**Comment by nzakas** [comment]
@idhard then unfortunately I don't have any other advice to offer. You're running up against a limitation of ESLint's current design and you don't have some flexibility in how you're calling ESLint, I'm afraid there's not an answer.

**Comment by fasttime** [comment]
> [@eslint/eslint-tsc](https://github.com/orgs/eslint/teams/eslint-tsc) shall we make `**/`-prefixed paths match all files regardless of base path?

My concern with this approach is that we'd be adding more complexity to pattern matching in config files without addressing some of the problems that will then arise. Like ignoring a particular file outside the base path.

I still think that adding a `basePath` property to config objects would be a more flexible solution to work with files in a different directory, and it wouldn't change the way other settings work.

**Comment by nzakas** [comment]
> Matching absolute paths might be confusing. For example, currently a pattern like `**/foo/` matches directories `foo` relative to the location of the config file (or `cwd` if the `--config` option was used). If the pattern `**/foo/` would be matched against absolute paths, then it would effectively ignore everything if a directory named `foo` appears anywhere in the ancestry.

@mdjermanovic You're talking about using this pattern in `ignores`, correct? In that case, they could change the pattern to `/**/foo/` to ensure it was matched only from the config file. As mentioned earlier, this is a breaking change, so there will be some edge cases we need to consider, and that's why I think we need to put this behind an unstable flag to start.

> I still think that adding a `basePath` property to config objects would be a more flexible solution to work with files in a different directory, and it wouldn't change the way other settings work.

@fasttime this discussion is not mutually exclusive to adding `basePath` (of which I'm in favor). As I stated previously, I don't see these as solving the same problem

**Comment by nzakas** [comment]
Another approach could be to add a new key to config objects, something like `matchAbsolute: true` that could change the behavior of `**/` patterns. (FWIW I'd prefer not to add a new key for this purpose.)

**Comment by winstliu** [comment]
I also ran into this when upgrading to flat config. We install & call ESLint from a centralized location, let's say `/build/tools/`. But the code I want to lint is in a completely different location, say `/src/`. This is just how our build system works and is not something I can change.

I would expect running eslint from the tools directory like `eslint --config /src/eslint.config.mjs /src/` would work as expected, but instead I am told that everything in src is ignored.

Right now I have to do something like 
```pwsh
# ESLint uses the current working directory to decide what to lint: https://github.com/eslint/eslint/issues/19118
pushd
cd /src
eslint
popd
```

Which feels quite clunky.

**Comment by nzakas** [comment]
@mdjermanovic @fasttime waiting for response to my last comment(s) here.

**Comment by fasttime** [comment]
I'm not concerned about explicitly marking the new behavior with a flag like `matchAbsolute: true`, but if we are going to add the ability to match all files outside the base path, I think it would be good to also provide a way to match only some of those files specifically, otherwise the new functionality would have very limited use cases.

Most, if not all, real-world projects have particular files that should be ignored, or treated differently from the rest, and it would be impractical to match those files in all directories using a pattern that starts with `**/`. See for example @mdjermanovic's last comment https://github.com/eslint/eslint/issues/19118#issuecomment-2624547136.

Parsing relative paths that start with `../` makes sense at this point, and I think we should add that ability if we also want to match all files with a prefix like `**/`.

**Comment by Tyoneb** [comment]
I just encountered the same issue while migrating from ESLint 7. 

Like mentioned in https://github.com/eslint/eslint/issues/19118#issuecomment-2641732464, I run ESLint as follows:
```
npx /tools/eslint/eslint --config /tools/eslint/config.mjs /some/source/folder
```

As indicated in the ESLint doc ([Specifying files and ignores](https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-and-ignores)) my issue comes from the fact that ESLint is now sensitive to the location from where it is run:
> If using an alternate config file via the `--config` command line option, then all patterns are evaluated relative to the current working directory.

The way I see it, being able to override the default base path with an additional command line flag would solve the issue, while remaining compatible with the current behavior:
```
npx /tools/eslint/eslint --config /tools/eslint/config.mjs --basePath /some/source/folder .
```

As for the rest of the conversation regarding the ability for config objects to match files outside of the base path, I find it very scary that `**/` could match absolute paths... I'm no expert here and I apologize if I'm overstepping, but I feel like providing an additional `basePath` property that could be either an absolute path or a relative one resolved from the overall base path would be a nice approach. And IMHO files glob pattern should always match paths ***relative*** to the config object base path.

**Comment by mdjermanovic** [comment]
> The way I see it, being able to override the default base path with an additional command line flag would solve the issue, while remaining compatible with the current behavior:
> 
> ```
> npx /tools/eslint/eslint --config /tools/eslint/config.mjs --basePath /some/source/folder .
> ```

I'm in favor of this approach, and it seems it would solve the use cases mentioned in the original post and other posts in this issue.

**Comment by nzakas** [comment]
I have no objections to adding a `--base-path` command line arg. This could be added onto the design for this:
https://github.com/eslint/eslint/issues/18948

If we don't want to change the interpretation of `**/`, I'm fine with moving forward just with the base path approach to see how far that gets us.

**Comment by fasttime** [comment]
> The way I see it, being able to override the default base path with an additional command line flag would solve the issue, while remaining compatible with the current behavior:
> ```
> npx /tools/eslint/eslint --config /tools/eslint/config.mjs --basePath /some/source/folder .
> ```

This would solve the problem of linting files outside the current directory. I'm only thinking that if a config file needs a base path to work as intended in the first place, it would be more convenient to have that information encoded in the config itself rather than passing it as an option through the command line. That way, even API consumers could use the new functionality without any changes.

**Comment by nzakas** [comment]
@fasttime that's why I'm saying this should be in additional to adding `basePath` as a key in the config object.

**Comment by Tyoneb** [comment]
I'm not sure I understand the necessity of the `basePath` property for a config object.

If the config object requires a base path to apply the glob patterns from, it can be either a relative path resolved from the "overall" base path, or an absolute path. 

In either case, the base path of the config object could be expressed directly in the `files` glob patterns. Sure, having an additional `basePath` property will reduce the length of the glob patterns but I feel that it would add complexity for little added value. Unless I'm missing something here?

**Comment by nzakas** [comment]
@Tyoneb see the issue around that: https://github.com/eslint/eslint/issues/18948

**Comment by Kristinita** [comment]
This issue is still relevant as of April 2025.

Thanks.

**Comment by stephen-carter-at-sf** [comment]
I just found the same problem in an attempt to switch to ESLint 9. I'm using the node api and calculateConfigForFile is simply returning undefined now when I supply any file that is not under my cwd. If in my testing, I switch to a relative file... even if it doesn't exist- then things pass again.

It took me forever to figure out what was going on. The behavior did indeed seem to change when going from legacy config to flat config.

That is, if my eslint options has cwd as '/some/path1' and I provide '/some/path1/dummy.js' then my "**/*.js" filter seems to be working. If I provide '/some/path2/dummy.js' then my filter seems to not be working anymore and it doesn't find this file. If I just provide 'dummy.js' then it finds it.

I assumed that **/*.js meant all files period.

Note that this is important to the Salesforce Code Analyzer because we wrap ESLint and allow users to be sitting in one directory like '/some/path1' (which serves as cwd) and may execute
```
sf code-analyzer run -r eslint -w '/some/path2' - w '/some/path3'
```
which essentially should scan both folders '/some/path2' and '/some/path3' while sitting in '/some/path1'. But now I don't know what I can set for the cwd to make things work.

@nzakas When can we expect some progress on this? Any workarounds we can do for now?

Otherwise, we are stuck with ESLint 8 still.

**Comment by nzakas** [comment]
@stephen-carter-at-sf all available updates on the issue. We don't have a timeline for addressing this as everyone is working in their spare time and there are a lot of issues to get through. If Salesforce would be interested in sponsoring development of this change, you have my email.

**Comment by stephen-carter-at-sf** [comment]
This is really causing a major issue for us. Can we at least introduce some mechanism to be able to specify that we want to match against any absolute file (not just the ones under the base directory)?

Is it intentional that you are not allowing a user to put a eslint config file in a random directory, cd into that directory and then call 
`npx eslint <someOtherDirectoryNotUnderCwd>`
?

For integration use cases where folks wrap ESLint in their own tools and have their own config... we have no way to invoke ESLint on other directories unless we change the cwd to the absolute parent folder of all the files that are to be scanned.

But again, this is an issue if a user supplies 2 files that do not live under the same root (E:/someFile1.js, D:/someFile2.js) for example.

Also changing cwd (the field in ESLint.Options) has other unintended consequences I believe besides just determining if isPathIgnored is true or false).

**Comment by nzakas** [comment]
As a note, it's possible that the [experimental config file resolution](https://eslint.org/docs/latest/use/configure/configuration-files#experimental-configuration-file-resolution) may work to solve some cases related to this error, as it broadens the scope of where `eslint.config.js` can live and be found. Definitely not saying it fixes everything, just that it adds a bit more flexibility.

**Comment by char101** [comment]
I think the problem is in these two lines

https://github.com/eslint/eslint/blob/b9080cf28d88f934941a545a033eb960eceeadbd/lib/eslint/eslint.js#L757-L761

`this.#configLoader.loadConfigArrayForFile` is aware of `overrideConfigFile` while `configs.getConfig` doesn't, it only looks up the config based on the file path.

I think if `overrideConfigFile` is given, then `configs.getConfig` should use that config rather than trying to find the config based on the file path.

For example, if I change it into this, then the input file is not ignored

```javascript
filePaths.map(async filePath => {
  const configs = await this.#configLoader.loadConfigArrayForFile(filePath);
  const { options } = privateMembers.get(this);
  const config = configs.getConfig(options.configFile);
```

**Comment by FabriceSalvaire** [comment]
For [Emacs/Flycheck](https://www.flycheck.org/en/latest/languages.html#javascript)  users, the current workaround is to add in `.emacs`:
```
(setq flycheck-eslint-args '("--flag" "unstable_config_lookup_from_file"))
```
Else, it reports anything.

The command run by Flycheck is: 
```
/home/.../node_modules/eslint/bin/eslint.js --flag unstable_config_lookup_from_file --format\=json --stdin --stdin-filename /home/.../index.js < /home/.../index.js
```

**Comment by stephen-carter-at-sf** [comment]
@nzakas Regarding [experimental config file resolution](https://eslint.org/docs/latest/use/configure/configuration-files#experimental-configuration-file-resolution), what is the Node Api equivalent to the unstable_config_lookup_from_file flag?

**Comment by nzakas** [comment]
We're still planning on addressing this, so not stale.

**Comment by Jaxo** [comment]
Well.  This is really a show-stopper.  My work-around (on Linux Ubuntu) is to create a directory INSIDE the base path, with hundreds of symbolic links to files that are OUTSIDE the base path.

**Comment by nzakas** [comment]
@Jaxo have you tried using `--flag v10_config_lookup_from_file`?

**Comment by slolo2000** [comment]
Hello @nzakas,

With this new flag, is it possible to handle cases like this:
I have a central "compilation" project using gruntjs that allows for JavaScript and CSS linting, compression, etc.

I have all my other projects at the same folder level as my "compilation" project. Ex:

> D:\Projects\Grunt
> D:\Projects\Project1
> D:\Projects\Project2
> D:\Projects\Projectx...

From the Grunt build project, I can compile any project with a command like: grunt --project=project1

My grunt configuration file contains the folders for each of my projects and accesses them with a path like: ../Project1/**/*.js

Ex:

```
eslint: {
	options: {
		overrideConfigFile: ".eslintrc.json",
		outputFile: "eslint-report-project.html",
		format: "html"
	},
	project1: {
		src: ["../Project1/scripts/**/*.js"]
	},
	project2: {
		src: ["../Project2/scripts/**/*.js"]
	},
	...
}
```

This folder configuration allows me to have only one instance of ESLINT and all the other plugins in one place, and I find it very convenient.

This worked very well with ESLINT V8 but no more in V9.

So, is the --flag v10_config_lookup_from_file can be useful for this type on projects configuration?

**Comment by nzakas** [comment]
Why don't you try it and see if it works for you?

**Comment by slolo2000** [comment]
I have done a test with files posted by @Peredery in the first post and the option `--flag v10_config_lookup_from_file` does not seems to change anything.

Please, have a look to the files in attachment and see if I have missed something:
[eslint.zip](https://github.com/user-attachments/files/21392880/eslint.zip)

example_8_X: `works great`
example_9_X: `0:0  warning  File ignored because outside of base path`

But, if I understand correctly what the option is supposed to do, it does not solve the problem of files that are outside the scope of the configuration file.
[https://eslint.org/docs/latest/use/configure/configuration-files#experimental-configuration-file-resolution](https://eslint.org/docs/latest/use/configure/configuration-files#experimental-configuration-file-resolution)

> ESLint will search for a configuration file by first starting in the directory of the file being linted and then searching up its ancestor directories until it finds a eslint.config.js file (or any other extension of configuration file). This behavior is better for monorepos, where each subdirectory may have its own configuration file.

In the test case, you only have **one config file** and potentially some .js files everywhere outside the main folder.

Hence the message: 

> 0:0  warning  File ignored because outside of base path

**Comment by nzakas** [comment]
@slolo2000 thanks for the feedback. We don't download random zip files off the internet for obvious reasons. :smile:

I don't have time to dig into this right now, which is why I was asking for folks to try the flag and see if it helps. 

We're still considering how to proceed on this.

**Comment by nzakas** [comment]
**TSC Summary:** Users reported that ESLint's flat config does not match files outside the current working directory, causing issues in monorepo and build tool setups. The discussion led to a proposal for a `--base-path` CLI option, allowing users to explicitly set the root directory for file matching. This approach was seen as a flexible, non-breaking solution that avoids confusing changes to glob pattern semantics.

**TSC Question:** Shall we move forward with this solution?

**Comment by sam3k** [comment]
The TSC group has agreed to move forward with the proposed `--base-path` option. The issue will be marked as accepted and "Waiting for RFC".

**Comment by saifulapm** [comment]
any update

**Comment by nzakas** [comment]
@saifulapm all updates are posted on the issue.

**Comment by fasttime** [comment]
I will open an RFC.
